### PR TITLE
docs: add "Fine-Tuning & Training LLMs" guide

### PR DESCRIPTION
**Pull request by saforem2**
## What

Adds a task-oriented **Fine-Tuning & Training LLMs** guide
(`docs/guides/llm-finetuning.md`) that ties together the three LLM
training/fine-tuning example modules, which today are only documented
one-per-module with no "which do I use / how do I scale" overview.

Covers:
- **The three paths** — `hf_trainer` (HF `Trainer`), `hf` (custom loop), and
  `fsdp_tp` (from-scratch FSDP2 + tensor/HSDP parallel) — with a "which path?"
  decision tip. Each links to its example page, API reference, and GitHub source.
- **Model targets** — HF paths: Llama-3.1-8B (default) / Qwen2.5 / Mistral /
  Llama-3.3-70B via `--model_name_or_path`; `fsdp_tp`: the `--model
  {s,m,l,xl,xxl,xxxl}` ladder + `agpt-2b`/`agpt-20b`, plus the
  arbitrary-HF-repo-id `--model` path (FSDP2-only, TP forced off).
- **Resource requirements** — per-target GPU/node + memory-lever tables
  (clearly labeled rough guidance, to be filled in from real runs).
- **Scaling levers** — `--tp`, `--dp-replicate`/`--dp-shard`,
  `--reshard-after-forward`, `--loss-impl`, `--act-mem-budget`.
- Runnable commands per path + a Perlmutter@NERSC walkthrough + a follow-up
  runs checklist.

## Wiring

- Added to the **📖 Guide** nav (next to Distributed Training) in `zensical.toml`.
- Pointer callout from `docs/examples/index.md`.

## Verification

- `zensical build` succeeds; **no new warnings** reference the new page.
- All intra-doc relative links (example pages, API refs, comparison page,
  distributed-training guide) verified to resolve to real files.

`release:skip` (docs only).

## Summary by Sourcery

Add a task-oriented guide for fine-tuning and training LLMs with ezpz and wire it into the documentation navigation and examples overview.

Documentation:
- Introduce a comprehensive "Fine-Tuning & Training LLMs" guide covering three training paths, model targets, scaling levers, resource guidance, and Perlmutter walkthrough.
- Link the new LLM fine-tuning guide from the examples index to help users choose between hf_trainer, hf, and fsdp_tp examples.
- Add the LLM fine-tuning guide to the main documentation navigation under the Guide section.

### CODE REVIEW

File: docs/guides/llm-finetuning.md [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        --model_name_or_path meta-llama/Llama-3.3-70B \
+        --fsdp=full_shard \
+        --block-size=4096 \
+```
+
+If you need `torch` and/or `mpi4py`, add the extras:
+
+```bash
+TMPDIR=$(pwd) uv run \
+    --with "git+https://github.com/saforem2/ezpz[torch,mpi4py]" \
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md]: Fixed in `cff0396` — the `uv` requirement now uses the declared extras: `ezpz[torch,mpi,hf]` (was `[torch,mpi4py]`). `hf` pulls in `accelerate` for the HF examples; `mpi` pulls `mpi4py`. Verified against `pyproject.toml` (`mpi = ["mpi4py"]`, `hf = ["accelerate"]`).

File: docs/guides/llm-finetuning.md [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        --streaming \
+        --dataset_name=eliplutchok/fineweb-small-sample \
+        --tokenizer_name meta-llama/Llama-3.1-8B \
+        --model_name_or_path meta-llama/Llama-3.1-8B \
+        --bf16=true \
+        --do_train=true \
+        --do_eval=true \
+        --report-to=wandb \
+        --logging-steps=1 \
+        --include-tokens-per-second=true \
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md]: Correct — verified `ezpz.examples.hf` does NOT normalize `include_tokens_per_second` (only `hf_trainer` does, via its rewrite pass). Fixed in `cff0396`: dropped `--include-tokens-per-second` from all Path 2 (`hf`) commands and kept `--include-num-input-tokens-seen`. Noted that `hf.py` reports `train/tokens_per_sec` itself, so no token-speed flag is needed there.

File: docs/guides/llm-finetuning.md [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        --include-num-input-tokens-seen=true \
+        --optim=adamw_torch \
+        --logging-first-step \
+        --include-for-metrics='inputs,loss' \
+        --max-eval-samples=100 \
+        --per_device_train_batch_size=1 \
+        --per_device_eval_batch_size=1 \
+        --block_size=8192 \
+        --gradient_checkpointing=true \
+        --fsdp=auto_wrap
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md]: Correct — verified `hf.py` only builds its FSDP plugin when `ACCELERATE_USE_FSDP=true` (`src/ezpz/examples/hf.py:278`) and never reads `training_args.fsdp`. Fixed in `cff0396`: the Path 2 commands now prefix `ACCELERATE_USE_FSDP=true` and drop the no-op `--fsdp=auto_wrap`, with an inline note explaining the difference from `hf_trainer`.

File: docs/guides/llm-finetuning.md [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+```bash
+TMPDIR=$(pwd) uv run \
+    --with "git+https://github.com/saforem2/ezpz" \
+    --python=$(which python3) \
+    ezpz launch python3 -m ezpz.examples.fsdp_tp \
+        --model xxl \
+        --tp 2 \
+        --dp-shard -1 \
+        --dp-replicate 1 \
+        --reshard-after-forward always \
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md:266]: This one is now valid: `--reshard-after-forward` landed on `main` with PR #192 (merged just before this fix). `fsdp_tp` accepts `--reshard-after-forward {always,never}` / `--no-reshard-after-forward` (the old `--sharding-strategy` is now a hidden deprecated alias). So the Path 3 snippet is correct as of current `main` — keeping it.

File: docs/guides/llm-finetuning.md [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+        --loss-impl loss-parallel \
+        --act-mem-budget 0.5 \
+        --compile \
+        --seq-len 4096 \
+        --batch-size 1 \
+        --report-to wandb
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md]: Correct — `fsdp_tp` has no `--report-to` (tracking is via `ezpz`'s `History` / `setup_wandb` + `WANDB_*` env, auto-enabled). Fixed in `cff0396`: removed `--report-to` from every `fsdp_tp` command (Path 3 headline, agpt-20b, arbitrary-HF, and Perlmutter) and added a note that wandb is auto-enabled there.

File: docs/guides/llm-finetuning.md [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+        --do_train=true \
+        --do_eval=true \
+        --report-to=wandb \
+        --logging-steps=1 \
+        --include-tokens-per-second=true \
+        --max-steps=100 \
+        --include-num-input-tokens-seen=true \
+        --optim=adamw_torch \
+        --logging-first-step \
+        --include-for-metrics='inputs,loss' \
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md:235]: Fixed in `cff0396` — dropped the `--include-*` Trainer flags from the Path 2 (`hf`) command; `hf.py` already logs `train/tokens_per_sec` itself, and `--include-tokens-per-second` isn't normalized there (fails on transformers v5). Kept `--include-num-input-tokens-seen` only.

File: docs/guides/llm-finetuning.md [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+them (wandb + the resource tables above):
+
+1. **Fit / OOM sweep per target.** Confirm the default configs (Llama-3.1-8B
+   on Paths 1 & 2; `--model xxl` on Path 3) actually fit at the stated GPU
+   counts; record peak memory and the smallest node count that fits. Adjust
+   `block_size` / `seq_len` / `act-mem-budget` and re-record on OOM.
+2. **Throughput / MFU.** Capture tokens/sec-per-GPU and MFU at steady state
+   (past step ~20, after any `torch.compile` warmup/recompile). HF paths:
+   `--include-tokens-per-second`. Path 3: the `tps_per_gpu` / `mfu` history
+   metrics.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md]: Clarified in `cff0396` — the follow-up "Throughput/MFU" item now spells out per path: Path 1 (`hf_trainer`) uses `--include-tokens-per-second`; Path 2 (`hf`) reports `train/tokens_per_sec` automatically (no flag); Path 3 (`fsdp_tp`) uses the `tps_per_gpu`/`mfu` history metrics.

File: docs/guides/llm-finetuning.md [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+    # Path 2: HF custom loop (same target)
+    ezpz launch python3 -m ezpz.examples.hf \
+        --dataset_name=eliplutchok/fineweb-small-sample --streaming \
+        --tokenizer_name meta-llama/Llama-3.1-8B \
+        --model_name_or_path meta-llama/Llama-3.1-8B \
+        --bf16=true --do_train=true --do_eval=true \
+        --report-to=wandb --logging-steps=1 --logging-first-step \
+        --include-tokens-per-second=true --include-num-input-tokens-seen=true \
+        --optim=adamw_torch --include-for-metrics='inputs,loss' \
+        --max-steps=100 --max-eval-samples=100 \
```
[the remark this answers is not in the document]
Comment (saforem2) [reply docs/guides/llm-finetuning.md:428]: Fixed in `cff0396` — the Perlmutter Path 2 command now uses `ACCELERATE_USE_FSDP=true` (instead of the ignored `--fsdp=auto_wrap`) and drops `--include-tokens-per-second`, matching the standalone Path 2 section.
